File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/request/BasicRequestEntity.java
package com.roxiemobile.networkingapi.network.rest.request;

import com.roxiemobile.androidcommons.diagnostics.Guard;
import com.roxiemobile.networkingapi.network.http.CookieStore;
import com.roxiemobile.networkingapi.network.http.HttpHeaders;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.net.URI;

public class BasicRequestEntity<T> implements RequestEntity<T> {

// MARK: - Construction

    protected BasicRequestEntity(@NotNull Builder<T> builder) {
        mUri = builder.mUri;
        mHttpHeaders = builder.mHttpHeaders;
        mCookieStore = builder.mCookieStore;
        mBody = builder.mBody;
    }

// MARK: - Properties

    @Override
    public @NotNull URI uri() {
        return mUri;
    }

    @Override
    public @Nullable HttpHeaders headers() {
        return mHttpHeaders;
    }

    @Override
    public @Nullable CookieStore cookieStore() {
        return mCookieStore;
    }

    @Override
    public @Nullable T body() {
        return mBody;
    }

// MARK: - Inner Types

    public static class Builder<T> {

        public Builder() {
            // Do nothing
        }

        public Builder(@NotNull RequestEntity<T> requestEntity) {
            Guard.notNull(requestEntity, "requestEntity is null");

            // Init instance variables
            mUri = requestEntity.uri();
            mHttpHeaders = requestEntity.headers();
            mCookieStore = requestEntity.cookieStore();
            mBody = requestEntity.body();
        }

        public <Ti> Builder(@NotNull RequestEntity<Ti> requestEntity, @Nullable T body) {
            Guard.notNull(requestEntity, "requestEntity is null");

            // Init instance variables
            mUri = requestEntity.uri();
            mHttpHeaders = requestEntity.headers();
            mCookieStore = requestEntity.cookieStore();
            mBody = body;
        }

        public @NotNull Builder<T> uri(@NotNull URI link) {
            mUri = link;
            return this;
        }

        public @NotNull Builder<T> headers(@Nullable HttpHeaders httpHeaders) {
            this.mHttpHeaders = httpHeaders;
            return this;
        }

        public @NotNull Builder<T> cookieStore(@Nullable CookieStore cookieStore) {
            mCookieStore = cookieStore;
            return this;
        }

        public @NotNull Builder<T> body(@Nullable T body) {
            mBody = body;
            return this;
        }

        public @NotNull RequestEntity<T> build() {
            Guard.notNull(uri(), "url is null");
            return new BasicRequestEntity<>(this);
        }

        protected @Nullable URI uri() {
            return mUri;
        }

        private @Nullable URI mUri;
        private @Nullable HttpHeaders mHttpHeaders;
        private @Nullable CookieStore mCookieStore;
        private @Nullable T mBody;
    }

// MARK: - Variables

    private final @NotNull URI mUri;

    private final @Nullable HttpHeaders mHttpHeaders;

    private final @Nullable CookieStore mCookieStore;

    private final @Nullable T mBody;
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/request/BasicRequestEntity.kt
package com.roxiemobile.networkingapi.network.rest.request

import com.roxiemobile.networkingapi.network.http.CookieStore
import com.roxiemobile.networkingapi.network.http.HttpHeaders
import com.roxiemobile.networkingapi.network.http.InMemoryCookieStore
import java.net.URI

open class BasicRequestEntity<TBody>: RequestEntity<TBody> {

// MARK: - Construction

    protected constructor(builder: Builder<TBody>) {
        this.link = requireNotNull(builder.link) { "link is null" }
        this.httpHeaders = builder.httpHeaders ?: HttpHeaders()
        this.cookieStore = builder.cookieStore ?: InMemoryCookieStore()
        this.body = builder.body
    }

// MARK: - Properties

    final override val link: URI

    final override val httpHeaders: HttpHeaders

    final override val cookieStore: CookieStore

    final override val body: TBody?

// MARK: - Methods

    override fun clone(): RequestEntity<TBody> {
        return Builder(this).build()
    }

// MARK: - Inner Types

    open class Builder<TBody> {

        constructor(requestEntity: RequestEntity<TBody>? = null) {
            this.link = requestEntity?.link
            this.httpHeaders = requestEntity?.httpHeaders
            this.cookieStore = requestEntity?.cookieStore
            this.body = requestEntity?.body
        }

        internal var link: URI? = null
            private set

        internal var httpHeaders: HttpHeaders? = null
            private set

        internal var cookieStore: CookieStore? = null
            private set

        internal var body: TBody? = null
            private set

        open fun link(link: URI): Builder<TBody> {
            this.link = link
            return this
        }

        open fun httpHeaders(httpHeaders: HttpHeaders): Builder<TBody> {
            this.httpHeaders = httpHeaders
            return this
        }

        open fun cookieStore(cookieStore: CookieStore): Builder<TBody> {
            this.cookieStore = cookieStore
            return this
        }

        open fun body(body: TBody?): Builder<TBody> {
            this.body = body
            return this
        }

        open fun build(): RequestEntity<TBody> {
            return BasicRequestEntity(this)
        }

        companion object {

            @JvmOverloads
            @JvmStatic
            fun <Ti, To> of(requestEntity: RequestEntity<Ti>, body: To? = null): Builder<To> {
                return Builder<To>().apply {
                    link(requestEntity.link)
                    httpHeaders(requestEntity.httpHeaders)
                    cookieStore(requestEntity.cookieStore)
                    body(body)
                }
            }
        }
    }

// MARK: - Companion

    companion object {

        @JvmOverloads
        @JvmStatic
        fun <Ti, To> of(requestEntity: RequestEntity<Ti>, body: To? = null): RequestEntity<To> {
            return Builder.of(requestEntity, body).build()
        }
    }
}
